data/transactions/logic/eval.go

```diff
@@ -432,6 +463,10 @@ func eval(program []byte, cx *evalContext) (pass bool, err error) {
 	if cx.stack[0].Bytes != nil {
 		return false, errors.New("stack finished with bytes not int")
 	}
+
+	// set side effects
+	cx.PastSideEffects[cx.GroupIndex].SetScratchSpace(cx.scratch)
```

Comment:
I think this is a wrong place to set data.
1. `prepareEvalParams` where PastSideEffects is allocated is called only for **app call* txn.
2. this `eval()` function is called for both **logic sig** and **app call**

This means: 1) logic sig txn would cause panic since `PastSideEffects` is empty  - please add a test before fixing 2)  values from **logic sig** are not available for reading in app call txns - is it intentional @jannotti @jdtzmn ?

Reply:
Would setting the data in the `EvalStateful` solve this?

Reply:
It depends on our design intent - if we want to completely ignore **logic sig** then yes. If we want **logic sig**'s scratch to be exposed to **app call** program then no.

Reply:
I think we should completely ignore logicsig.  Two reasons a) The rule is that you can only see calls that are "earlier" in the group.  That rule does not make sense for logicsigs, as they get evaluated first.  However, the rule would now be quite weird. b) The bigger reason is: what if an app call is logsigged?  Which state do you see?  I think it's much cleaner to say that `gload` is only for accessing app state.

Reply:
@algorandskiy I've added a check so that past side effects are only set from within an application call.

Reply:
why not moving it into EvalStateful?